### ISSUE TITLE
Picking up large quantities of items takes too long

### ISSUE THREAD

**Issue by anoobindisguise** [CleverRaven/Cataclysm-DDA#58793]
### Is your feature request related to a problem? Please describe.

This is a bit of a general issue, but picking up lots of items is very slow and painful. Especially now with the new tailoring system, looting tailoring basements in particular is a very slow process. But if I for example cut up a turnout coat into patches, it takes something like 40 minutes to pick up 700 patches given that each patch takes about 4 seconds to put into a backpack. This can be partially alleviated by not cutting up clothing until the last moment, and also by putting sheets/patches into plastic bags, garbage bags, canvas sacks, etc, in order to reduce the further number of item transactions into a single unit, but none of these address the root problem.

The main issue is that you have to pay the move cost of each container per each item that is moved from it. This gets even worse with sub-containers, which is very noticeable with the new bugout bags. Let's say I kill a zombie, and inside his bugout bag is a can of coffee powder, an IFAK, a gun and some clips and bullets, a rehydration drink, and some other things I don't care about. So I want to loot a lot of those, but it takes multiple minutes to do so, especially since I don't want everything inside the IFAK or the IFAK itself so the costs to remove [Bandages, Medical Gauze, Hemostatic Powder, Adhesive Bandages] are compounded for both the IFAK and the backpack itself. Then I have to pay all the costs to put them into my own containers, so it ends up being like 8-9 seconds per item.

### Solution you would like.

There needs to be some sort of sane way to speed this process up. There could be diminishing returns on time needed to pick up loose items, to represent you grabbing a whole handful or clump of whatever and then moving it all at once - you could hold dozens of the aforementioned patches at once for example. Additionally, when moving many items between containers, the cost could be paid just once in the whole process rather than each step of the way. You unzip the bugout bag, unzip your backpack, and then move items between them one by one. Currently it's "you unzip the bugout bag, take the item, zip the bugout bag, unzip your backpack, put in the item, zip your backpack, repeat"

### Describe alternatives you have considered.

Unsure but currently it's kind of insanity and the "bundle of x" implementation is really hacky and not great IMO.

**Comment by mqrause** [comment]
Yesterday I discovered things like this are strongly connected to UI redrawing. Turning off the "force redraw" option in `options > graphics` should speed up activities like this a lot.

**Comment by anoobindisguise** [comment]
> Yesterday I discovered things like this are strongly connected to UI redrawing. Turning off the "force redraw" option in `options > graphics` should speed up activities like this a lot.

Does this affect the real in-game time too? Or just the time spent waiting for each pickup in terms of real time? The 40 minutes I referred to was in-game-time, 30-60 seconds IRL.

**Comment by mqrause** [comment]
Oh, sorry, I'm only half awake and only read half the issue and was already half thinking about writing an issue about my findings, so I came to a premature conclusion. No, ingame time is not affected by that. It's purely (lots of) extra processing time spent on redrawing the UI that's slowing down the game in general.

**Comment by Zireael07** [comment]
I think we already have an issue for "lots of items take lots of in-game time". Key words: rags, thread, bullets

**Comment by descan** [comment]
Would it be possible to address part of this issue by having bags and containers have two states - closed, and open? With the open state having a *much* lower item move speed, at the cost of not being as secure (things dropping out as you move, so you don't run around with an open bag because it's faster). 

Relatedly to this issue, I would LOVE an option to do the real-life equivalent of just grabbing a bag and upending it onto the floor. I put a lot of stuff into bags to make hauling them around when I don't have a vehicle faster both IRL and in-game, but when it comes time to actually sorting them out, getting the items out of the bag is a bit of a pain. Unloading dumps it into your inventory (meaning you have to deal with sorting it twice, or taking off all your gear to force it to drop to the ground instead) and the advanced inventory screen doesn't let you take items from within bags - it only allows the top-level of nested containers to be moved.

**Comment by mqrause** [comment]
> Relatedly to this issue, I would LOVE an option to do the real-life equivalent of just grabbing a bag and upending it onto the floor. I put a lot of stuff into bags to make hauling them around when I don't have a vehicle faster both IRL and in-game, but when it comes time to actually sorting them out, getting the items out of the bag is a bit of a pain.

There is an `unload container` action that is unbound by default. You can dump the contents of containers on a nearby tile with that. Dunno about ingame time cost, but I assume it has the same issues as regular dropping.

**Comment by chrispikula** [comment]
Also, things like your hand having anything in it, and encumbrance, also play parts in slowing down item transfer.

But, yeah, there really should be a 'shovel all this shit in here' option, at the expense of a much worse packing number or something.

**Comment by deafmole** [comment]
I think it would help to change the status of clothing 'Spare Parts' into something else, or at least add Ammo-like properties to it. I vaguely remember that they have been classified as Ammo in the past which led to confusion. 
What I mean by that is the difference in the Advanced Inventory Manager Menu ('/') of something like Ammo or Cigarettes(Drugs). Another (similar) thing could be automatically turning small 'Spare Parts' objects in the same container automatically into a batch of them (20-50) via maybe a fake nested container. I'm not sure how the batch time reduction in crafting gets calculated (of 1 second recipees theoretically get a timereduction), but that could be another implementationoption for moving/picking up/dropping small objects and Spare Parts.

**Comment by mqrause** [comment]
Items that are `count_by_charges` behave as if they're just a single item. You can handle 10 tons of steel mostly the same as 100 grams. That's not something we want, which is also why cotton sheets and similar items don't do that anymore and why that whole mechanic will be removed in the future.

As the OP mentioned, the main issue is that currently the cost for accessing the container is added for every single item you drop. Which is probably not too hard to fix. The activity just has to remember which containers were accessed already and only count that for the first item taken from or put there.

**Comment by PatrikLundell** [comment]
@mqrause posted while I wrote, so there's overlap...

Movement of multiple items should probably be a two pass affair where movement of an item opens up the container(s) in needs to pass through and deposits the item, but does not close the container(s). The second item that's moved and onward would then skip the time for opening any container(s) that's been left open, and the whole process would have a second pass to close up the containers that had been opened.

Interruptions should leave the containers open, spilling items all over the place when moving. This requirement would mean that the game would need to actually have an open container state, rather than having it as a purely abstract state kept track of internally in the movement logic. However, I think the internal state implementation may be good enough for a first implementation (meaning that all containers automatically close in no time on an interrupt).

The problem with shifting from "single item with any number of charges" to "stacks of individual items" is that the latter doesn't have any logic for moving more than one item at a time. If I have a pile of coins or sheets of fabric, I don't carefully more them one by one into my open sack (or whatever destination I have that doesn't add opening/closing time), but rather grab a lot of them in each action. Cleaning out a sewing basement wouldn't take 2-3 hours in the real world, but it does in the game because of the inability to move sub stacks of items.

What about adding a movement stack value to items that don't have charges (with the default being one, i.e. single item)? When trying to move stack of 100 sheets of fabric, a movement stack size of 10 would have the game trying to move the bulk of up to 10 sheets at a time into spaces that can take that volume/weight until no stacks can be moved, at which time individual items would be moved (to fill out space too small for a whole stack).

**Comment by deafmole** [comment]
@mqrause I absolutely get that, which is why I mentioned fake nested containers - so that different items will get respective batch-properties that make sense.

**Comment by Zireael07** [comment]
@deafmole: What you propose ("handling like ammo") is basically "count_by_charges hack" - it was attempted once and rejected as a , well, hack.

**Comment by deafmole** [comment]
@Zireael07  Oh no, don't get me wrong, I am against the charges-idea in general. 
I'm for handling of stacks/batches of a certain item, which would happen automatically, as stated above. 
The reason why I mentioned the charges-'solution' is that while it still makes no sense, it is at least **bearable** as player. Currently I am literally walking away from the game when picking up craftingmaterials, because it takes 1-2 minutes to pick stuff up.

**Comment by PatrikLundell** [comment]
I don't think you should lose everything immediately just because "containers" are open, just that there should be a risk for it. I've also realized that this might be a thing in favor of carried containers rather than pockets, as carried containers probably would have to be placed on the ground before you can stuff things into them efficiently. This would also mean that a backpack on the ground would risk being left behind, but would also not contribute to your encumbrance when you're interrupted.

**Comment by Zireael07** [comment]
What's the state with the recent inventory performance PRs?

**Comment by mqrause** [comment]
I don't think they do anything about this, it's quite a different issue.

**Comment by anoobindisguise** [comment]
this was improved a long time ago, probably doesn't need to be open still
